plutus-core/plutus-core/src/PlutusCore/Builtin/KnownType.hs

```diff
@@ -195,12 +193,74 @@ a ton of instances (and add new ones whenever we need them), wrap and unwrap all
 
 -- See Note [MakeKnownM and ReadKnownM being type synonyms].
 -- | The monad that 'makeKnown' runs in.
-type MakeKnownM = ExceptT KnownTypeError Emitter
+data MakeKnownM a
+    = MakeKnownFailure !(DList Text) !KnownTypeError
+    | MakeKnownSuccess !a
+    | MakeKnownSuccessWithLogs !(DList Text) !a
+
+withEmitMakeKnownM :: DList Text -> MakeKnownM a -> MakeKnownM a
+withEmitMakeKnownM logs1 (MakeKnownSuccessWithLogs logs2 x) =
+    MakeKnownSuccessWithLogs (logs1 <> logs2) x
+withEmitMakeKnownM logs1 (MakeKnownSuccess x) =
+    MakeKnownSuccessWithLogs logs1 x
+withEmitMakeKnownM logs1 (MakeKnownFailure logs2 err) =
+    MakeKnownFailure (logs1 <> logs2) err
+{-# INLINE withEmitMakeKnownM #-}
+
+instance Functor MakeKnownM where
+    fmap _ (MakeKnownFailure logs err)       = MakeKnownFailure logs err
+    fmap f (MakeKnownSuccess a)              = MakeKnownSuccess (f a)
+    fmap f (MakeKnownSuccessWithLogs logs a) = MakeKnownSuccessWithLogs logs (f a)
+    {-# INLINE fmap #-}
```

Comment:
All these `INLINE` annotations seem a little risky to me. If you have any substantial `do` block in `MakeKnownM` you're going to be at risk of exponential code bloat. Maybe we should just do `INLINABLE`?

Reply:
Why exponential? It's a constant factor: every call to `>>=` becomes, what, six pattern matches. Not too much of a deal, given that we don't do much in `MakeKnownM` and we want all of it to be efficient, apart from stuff in tests, but it's still not that much.

Reply:
Yes, you're right, it's not exponential. It would only become so if GHC got happy with case-of-case, which it's probably too smart to do. I'd be interested to know if performance is actually worse with `INLINABLE`.

Reply:
> Yes, you're right, it's not exponential. It would only become so if GHC got happy with case-of-case, which it's probably too smart to do. 

Hm, there's something to worry about here indeed. So let's say we didn't have `MakeKnownSuccessWithLogs`, then

```haskell
(a >>= f) >>= g
```

would expand into

```haskell
case (case a of
        MakeKnownFailure logs err -> MakeKnownFailure logs err
        MakeKnownSuccess x        -> f x) of
    MakeKnownFailure logs err  -> MakeKnownFailure logs err
    MakeKnownSuccess y         -> g y
```

which would be turned by `case-of-case` into

```haskell
case a of
    MakeKnownFailure logs err ->
        case MakeKnownFailure logs err of
            MakeKnownFailure logs err  -> MakeKnownFailure logs err
            MakeKnownSuccess y         -> g y
    MakeKnownSuccess x        ->
        case f x of
            MakeKnownFailure logs err  -> MakeKnownFailure logs err
            MakeKnownSuccess y         -> g y
```

which would be turned by `case-of-known-constructor` into

```haskell
case a of
    MakeKnownFailure logs err ->
        MakeKnownFailure logs err
    MakeKnownSuccess x        ->
        case f x of
            MakeKnownFailure logs err  -> MakeKnownFailure logs err
            MakeKnownSuccess y         -> g y
```

which is still perfectly linear.

But we do have `MakeKnownSuccessWithLogs` and so it is indeed very much possible we'll get duplication in the two success cases. 

But then

1. we do want the duplication, since it's important for this stuff to be as efficient as possible
2. we won't actually get it, because our `f` and `g` are going to be known and so the `case-of-known-constructor` will save us

Really, I don't think we should worry about this. Especially compared to the ridiculous code bloats with meanings of builtins inlining all the way through the pipeline at each and every step until the very construction of `EvaluationContext` (we won't have any problems with `MakeKnownM` at any step, because the denotations have always been perfectly linear due to everything being fully known).

> I'd be interested to know if performance is actually worse with `INLINABLE`.

It probably isn't any worse, but we're building quite a house of cards with all the optimizations here, I don't want to make it any more shaky unless we have to.